File: forwarder/app/src/main/java/com/codey/forwarder/CodeyPairingLink.java
package com.codey.app;

import java.net.URI;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Map;

final class CodeyPairingLink {
    static final String SCHEME = "codey";
    static final String HOST = "pair";

    final String baseUrl;
    final String deviceCode;
    final String userCode;

    private CodeyPairingLink(String baseUrl, String deviceCode, String userCode) {
        this.baseUrl = baseUrl;
        this.deviceCode = deviceCode;
        this.userCode = userCode;
    }

    static CodeyPairingLink parse(String value) throws Exception {
        if (value == null || value.trim().isEmpty()) {
            throw new Exception("Pairing link is empty.");
        }

        URI uri = new URI(value.trim());
        if (!SCHEME.equalsIgnoreCase(uri.getScheme()) || !HOST.equalsIgnoreCase(uri.getHost())) {
            throw new Exception("QR code is not a Codey pairing link.");
        }

        Map<String, String> query = parseQuery(uri.getRawQuery());
        String baseUrl = trim(query.get("baseUrl"));
        String deviceCode = trim(query.get("deviceCode"));
        String userCode = trim(query.get("userCode"));
        if (baseUrl.isEmpty() || deviceCode.isEmpty() || userCode.isEmpty()) {
            throw new Exception("Pairing link is missing required fields.");
        }

        return new CodeyPairingLink(baseUrl, deviceCode, userCode);
    }

    String approvalUrl() {
        String normalizedBaseUrl = baseUrl;
        while (normalizedBaseUrl.endsWith("/")) {
            normalizedBaseUrl = normalizedBaseUrl.substring(0, normalizedBaseUrl.length() - 1);
        }
        return normalizedBaseUrl + "/device?userCode=" + urlEncode(userCode);
    }

    private static Map<String, String> parseQuery(String rawQuery) throws Exception {
        Map<String, String> query = new HashMap<>();
        if (rawQuery == null || rawQuery.trim().isEmpty()) {
            return query;
        }

        String[] pairs = rawQuery.split("&");
        for (String pair : pairs) {
            int separator = pair.indexOf('=');
            String key = separator >= 0 ? pair.substring(0, separator) : pair;
            String value = separator >= 0 ? pair.substring(separator + 1) : "";
            query.put(urlDecode(key), urlDecode(value));
        }
        return query;
    }

    private static String urlDecode(String value) throws Exception {
        return URLDecoder.decode(value, "UTF-8");
    }

    private static String urlEncode(String value) {
        try {
            return URLEncoder.encode(value, "UTF-8");
        } catch (Exception ignored) {
            return value;
        }
    }

    private static String trim(String value) {
        return value == null ? "" : value.trim();
    }
}

File: forwarder/app/src/main/java/com/codey/forwarder/CodeyPairingLink.kt
package com.codey.app

import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder

data class CodeyPairingLink(
    val baseUrl: String,
    val deviceCode: String,
    val userCode: String
) {
    fun approvalUrl(): String {
        var normalizedBaseUrl = baseUrl
        while (normalizedBaseUrl.endsWith("/")) {
            normalizedBaseUrl = normalizedBaseUrl.dropLast(1)
        }
        return "$normalizedBaseUrl/device?userCode=${urlEncode(userCode)}"
    }

    companion object {
        private const val SCHEME = "codey"
        private const val HOST = "pair"

        fun parse(value: String?): CodeyPairingLink {
            if (value.isNullOrBlank()) {
                throw Exception("Pairing link is empty.")
            }

            val uri = URI(value.trim())
            if (!SCHEME.equals(uri.scheme, ignoreCase = true) ||
                !HOST.equals(uri.host, ignoreCase = true)
            ) {
                throw Exception("QR code is not a Codey pairing link.")
            }

            val query = parseQuery(uri.rawQuery)
            val baseUrl = query["baseUrl"].trimOrEmpty()
            val deviceCode = query["deviceCode"].trimOrEmpty()
            val userCode = query["userCode"].trimOrEmpty()
            if (baseUrl.isEmpty() || deviceCode.isEmpty() || userCode.isEmpty()) {
                throw Exception("Pairing link is missing required fields.")
            }

            return CodeyPairingLink(baseUrl, deviceCode, userCode)
        }

        private fun parseQuery(rawQuery: String?): Map<String, String> {
            if (rawQuery.isNullOrBlank()) {
                return emptyMap()
            }

            return rawQuery.split("&").associate { pair ->
                val separator = pair.indexOf('=')
                val key = if (separator >= 0) pair.substring(0, separator) else pair
                val value = if (separator >= 0) pair.substring(separator + 1) else ""
                urlDecode(key) to urlDecode(value)
            }
        }

        private fun urlDecode(value: String): String {
            return URLDecoder.decode(value, "UTF-8")
        }

        private fun urlEncode(value: String): String {
            return try {
                URLEncoder.encode(value, "UTF-8")
            } catch (_: Exception) {
                value
            }
        }
    }
}
